File: potion.cpp
/**
 * @author : Rushikesh Keshavrao Deshmukh
 * @email  : rushissv@gmail.com
 * @date   : 10-Feb-2024

This program is written to solve the following problem:
Consider a potion(recipe) with N ingredients.
The objective is to determine the proportions of all the N ingredients in the potion.
Input consists of (N-1) lines. Each line has four integer values as follows:

	  index_i, index_j, amount_i, amount_j

index_i denotes the index of ingredient i
index_j denotes the index of ingredient j
amount_i denotes the relative proportionate amount of ingredient i
amount_j denotes the relative proportionate amount of ingredient j

Where
0 <= i, j < N
0 < amount_i, amount_j < 10
2 < N < 200

Above four integer values indicate the relative proportions of the two ingredients i, j.
Like this (N-1) inputs are given for different pair of ingredients.

The objective is to determine the proportions of all the N ingredients in potion.
If unique solution exists, then, the output for proportions of all N ingredients should add to 1.0

The inputs may be invalid or are not sufficient to determine the unique solution.
In such cases, the program should report that unique solution does not exist.
Otherwise program should output the proportion values for all N ingredients.

If solution exists, the final proportion values of all the N ingredients must satisfy all the (N-1) relative proportions given in the inputs.

Sample input:

3
0 1 9 4
2 1 1 2

First line in the input (3 in this case) indicates the total number of ingredients in potion.
After the first line, (N-1) lines follow. In this case, (3-1) = 2 lines follow.
Each line that follows specifies the relative proportion of two ingredients.
Second line mentioned above specifies that the relative proportion of ingredients 0, 1 must be 9:4
Similarly, last line specifies that the relative proportion of ingredients 2, 1 must be 1:2

The answer to above problem is as follows:
(9/15)    (4/15)    (2/15)
0.6000    0.2667    0.1333

Above final proportion values satisfy all the relative proportion values mentioned in the inputs.

Solution Approach: Incremental formation & merging of disjoint sets for ingredients.

Assume the mathematical inductive step k where the existing proportions of all the ingredients in potion are properly defined.
These ingredients may belong to one or multiple disjoint groups.
The potion may contain some ingredients for which the proportions are not yet defined and they don't belong to any group.

Now we need to add new proprtion entry for two ingredients.

Following are the various possibilities:

1. Both the ingredients don't belong to any group.
   Form a new group and add both ingredients to this new group.

2. One of the ingredients don't belong to any group and the other belongs to existing group.
   Add the ingredient with missing group to the existing group.
   Modify the proportion values to match all the proportion constraints.

3. Both ingredients belong to already existing different groups.
   Modify the proportion values for both the groups to match all the proportion constraints.
   Merge both these groups.

4. Both ingredients belong to already existing same group.
   Invalid input. This leads to disjoint groups because we only have (N-1) entries and N ingredients.

At the end if the the total number of unique groups in potion is not equal to one, report that no solution exists.
Otherwise determine the proportions for each ingredient in potion based on its amount in potion.

This approach is implemented in determineProportions() function in this file.
    Time complexity  = O(N^2)
    Space complexity = O(N)
*/
#include <algorithm>
#include <assert.h>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Proportion
{
	struct Ingredient
	{
		int index;
		int amount;
	};

	Ingredient one;
	Ingredient two;
};

static istream& operator >> (istream& is, Proportion& proportion)
{
	is >> proportion.one.index;
	is >> proportion.two.index;
	is >> proportion.one.amount;
	is >> proportion.two.amount;

	return is;
}

class Potion
{
	struct Ingredient
	{
		int amount;
		int group;
	};

	vector<Ingredient> ingredients;
	int maxGroup = 1;

public:

	Potion(const int N) : ingredients(N) {}

	bool addProportion(const Proportion& iProportion)
	{
		const Proportion::Ingredient one = iProportion.one;
		const Proportion::Ingredient two = iProportion.two;

		if (!isGroupAssigned(one.index) && !isGroupAssigned(two.index))
		{
			addNewGroup(iProportion);
			return true;
		}

		bool oneAssigned = false;
		Proportion::Ingredient assigned;
		Proportion::Ingredient unassigned;

		if (isGroupAssigned(one.index) && !isGroupAssigned(two.index))
		{
			oneAssigned = true;
			assigned = one;
			unassigned = two;
		}
		else if (!isGroupAssigned(one.index) && isGroupAssigned(two.index))
		{
			oneAssigned = true;
			assigned = two;
			unassigned = one;
		}

		if (oneAssigned)
		{
			int existingAssignedAmount = getAmount(assigned.index);
			multiply(assigned.amount, getGroup(assigned.index));
			set({ unassigned.index, unassigned.amount * existingAssignedAmount }, getGroup(assigned.index));
			return true;
		}

		if (isGroupAssigned(one.index) && isGroupAssigned(two.index))
		{
			if (areGroupsSame(one.index, two.index))
				return false;

			const int value1 = getAmount(one.index);
			const int value2 = getAmount(two.index);

			multiply(one.amount * value2, getGroup(one.index));
			multiply(two.amount * value1, getGroup(two.index));
			modifyGroup(getGroup(two.index), getGroup(one.index));

			return true;
		}

		return false;
	}

	bool isSingleGroup() const
	{
		const int group = ingredients[0].group;
		return std::all_of(ingredients.begin(), ingredients.end(), [&group](const Ingredient & ingredient)-> bool {
			return ingredient.group == group; });
	}

	void getProportions(vector<double>& oProportions) const
	{
		assert(oProportions.size() == ingredients.size());
		const int N = oProportions.size();

		int sm = 0;
		for (int i = 0; i < N; ++i)
			sm += ingredients[i].amount;

		for (int i = 0; i < N; ++i)
			oProportions[i] = (double)ingredients[i].amount / sm;
	}

private:

	bool isGroupAssigned(const int index) const
	{
		return ingredients[index].group != 0;
	}

	int getAmount(const int index) const
	{
		return ingredients[index].amount;
	}

	int getGroup(const int index) const
	{
		return ingredients[index].group;
	}

	void setGroup(const int index, const int group)
	{
		ingredients[index].group = group;
	}

	void set(const Proportion::Ingredient & ingredient, const int group)
	{
		ingredients[ingredient.index].amount = ingredient.amount;
		ingredients[ingredient.index].group = group;
	}

	void addNewGroup(const Proportion & proportion)
	{
		set(proportion.one, maxGroup);
		set(proportion.two, maxGroup);
		++maxGroup;
	}

	bool areGroupsSame(const int index1, const int index2) const
	{
		return ingredients[index1].group == ingredients[index2].group;
	}

	void multiply(const int m, const int group)
	{
		for (auto& ingredient : ingredients)
		{
			if (group != ingredient.group)
				continue;
			ingredient.amount *= m;
		}
	}

	void modifyGroup(const int fromGroup, const int toGroup)
	{
		for (auto& ingredient : ingredients)
		{
			if (ingredient.group == fromGroup)
				ingredient.group = toGroup;
		}
	}
};

static bool determineProportions
(
	const int N,
	const vector<Proportion> & iProportions,
	vector<double> & oProportions
)
{
	assert(iProportions.size() == (oProportions.size() - 1));
	assert(N == (int)oProportions.size());

	Potion potion(N);
	for (int i = 0; i < (N - 1); ++i)
	{
		if (!potion.addProportion(iProportions[i]))
			return false;
	}

	if (!potion.isSingleGroup())
		return false;

	potion.getProportions(oProportions);
	return true;
}

int main()
{
	int N;
	cin >> N;

	vector<Proportion> iProportions(N - 1);

	for (int i = 0; i < (N - 1); ++i)
		cin >> iProportions[i];

	vector<double> oProportions(N);
	const bool success = determineProportions(N, iProportions, oProportions);

	if (!success)
	{
		cout << "No solution exists." << endl;
		return 0;
	}

	cout << "Proportions = \n";
	for (int i = 0; i < N; ++i)
		cout << oProportions[i] << " ";
	cout << endl;

	return 0;
}

File: Potion.java
/*
 * Copyright (c) 2024.
 * @author : Rushikesh Keshavrao Deshmukh
 * @email: rushissv@gmail.com
 *
 * This file contains program to solve problem to determine proportions for a potion (recipe).
 * The potion has n ingredients.
 * (n-1) inputs for the relative proportions of two ingredients are given as follows:
 *     index1, index2, amount1, amount2
 *
 *     index1 denotes the index of ingredient 1
 *     index2 denotes the index of ingredient 2
 *     amount1 denotes the amount of ingredient 1
 *     amount2 denotes the amount of ingredient 2
 *
 * Determine the proportions of all the ingredients if unique solution exists.
 * The proportions of all ingredients should add to 1.0 if solution exists
 * If solution does not exist or if there is no unique solution, all the proportions should be zero.
 *
 * The program should be able to determine if unique solution exists or not.

Sample input:

3
0 1 9 4
2 1 1 2

3 value mentioned in first line is the total number of ingredients in potion.
Second line indicates that the relative proportions of ingredients 0, 1 should be 9:4
Similarly last line indicates that the relative proportions of ingredients 2, 1 should be 1:2
The final proportions of all ingredients should respect all the given relative proportions if solution exists.
The answer to above problem is as follows:
(9/15)    (4/15)    (2/15)
0.6000    0.2667    0.1333

Constraints:
2 < N < 200
0 < P < 10

Where N is the number of ingredients in potion.
P is the proportion amount for the ingredient specified in the inputs.


* Approach : Depth first search traversal (DFS) algorithm on the graph of ingradients.
This graph has ingredients as nodes and relative proportions given in input as edges between the nodes.
Since there are N nodes and (N-1) edges, following properties can be deduced:
  - For valid inputs, it must form a tree without cycles.
  - For valid inputs, all the nodes must be connected.
If above conditions are satisfied, then, problem can be solved and has unique solution.
Otherwise, unique solution cannot be found.

To perform depth first traversal, connections data-structure needs to be built using the inputs.
This DFS algorithm is implemented using determineProportionsUsingDFS() function in this file.
 */

package potion;

import java.text.NumberFormat;
import java.util.*;

public class Potion
{
    public static void main(String[] args)
    {
        List<Proportion> iProportions = readInputs();
        final int N = iProportions.size() + 1;

        Optional<List<Double>> oProportions = determineProportionsUsingDFS(N, iProportions);
        if (oProportions.isEmpty())
        {
            System.out.println("No solution exists.");
            return;
        }

        printSolution(oProportions.get());
    }

    private static Optional<List<Double>> determineProportionsUsingDFS(final int N, final List<Proportion> iProportions)
    {
        assert iProportions.size() == (N - 1);

        List<Double> proportions = Arrays.asList(new Double[N]);
        Collections.fill(proportions, 0.0);

        List<List<Integer>> connections = new ArrayList<>(N);
        for (int i = 0; i < N; ++i)
            connections.add(new ArrayList<>());

        for (int i = 0; i < (N - 1); ++i)
        {
            connections.get(iProportions.get(i).one.index).add(i);
            connections.get(iProportions.get(i).two.index).add(i);
        }

        List<Ingredient> ingredients = new ArrayList<>(N);
        ingredients.add(new Ingredient(0, 1));

        List<Boolean> visited = Arrays.asList(new Boolean[N]);
        Collections.fill(visited, false);
        visited.set(0, true);

        Queue<Integer> q = new ArrayDeque<>(N);
        q.add(0);
        int count = 1;

        while (!q.isEmpty())
        {
            final int index = q.remove();

            for (int i = 0; i < connections.get(index).size(); ++i)
            {
                final Proportion proportion = iProportions.get(connections.get(index).get(i));
                final Ingredient current = (proportion.one.index == index) ? proportion.one : proportion.two;

                assert current.index == index;

                final Ingredient adj = (proportion.one.index == index) ? proportion.two : proportion.one;

                if (visited.get(adj.index))
                    continue;

                visited.set(adj.index, true);
                q.add(adj.index);

                int currentAmount = -1;

                for (int j = 0; j < ingredients.size(); ++j)
                {
                    if(ingredients.get(j).index == index)
                        currentAmount = ingredients.get(j).amount;

                    ingredients.get(j).amount *= current.amount;
                }

                assert currentAmount > 0;
                
                ingredients.add(new Ingredient(adj.index, adj.amount * currentAmount));

                ++count;
            }
        }

        if (count != N)
            return Optional.empty();

        assert ingredients.size() == N;

        Collections.sort(ingredients, (ingredient1, ingredient2) -> ingredient1.index - ingredient2.index);
        for (int i = 0; i < ingredients.size(); ++i)
        {
            if (ingredients.get(i).index != i)
                return Optional.empty();
        }

        int sum = 0;
        for (int i = 0; i < ingredients.size(); ++i)
            sum += ingredients.get(i).amount;

        for (int i = 0; i < ingredients.size(); ++i)
            proportions.set(i, (double) ingredients.get(i).amount / sum);

        return Optional.of(proportions);
    }

    private static void printSolution(List<Double> proportions)
    {
        NumberFormat nf = NumberFormat.getInstance();
        nf.setMaximumFractionDigits(3);

        System.out.println();
        System.out.print("Proportions = [");

        for (int i = 0; i < (proportions.size() - 1); ++i)
            System.out.print(nf.format(proportions.get(i)) + ", ");

        System.out.print(nf.format(proportions.get(proportions.size() - 1)));
        System.out.println("]");
    }

    private static List<Proportion> readInputs()
    {
        Scanner scanner = new Scanner(System.in);
        final int N = scanner.nextInt();

        List<Proportion> iProportions = new ArrayList<>(N - 1);

        for (int i = 0; i < (N - 1); ++i)
        {
            final int index1 = scanner.nextInt();
            final int index2 = scanner.nextInt();
            final int amount1 = scanner.nextInt();
            final int amount2 = scanner.nextInt();

            Proportion proportion = new Proportion(index1, index2, amount1, amount2);
            iProportions.add(proportion);
        }

        return iProportions;
    }
}

class Ingredient
{
    Ingredient(int iIndex, int iAmount)
    {
        this.index = iIndex;
        this.amount = iAmount;
    }

    int index;
    int amount;
}

class Proportion
{
    Proportion(int index1, int index2, int amount1, int amount2)
    {
        one = new Ingredient(index1, amount1);
        two = new Ingredient(index2, amount2);
    }

    Ingredient one;
    Ingredient two;
}
